src/script/sign.h

```diff
@@ -68,6 +68,392 @@ struct SignatureData {
+static const uint8_t PSBT_SEPARATOR = 0x00;
+
+template<typename X>
+std::vector<unsigned char> SerializeToVector(uint8_t header, const X& x)
+{
+    std::vector<unsigned char> ret;
+    CVectorWriter ss(SER_NETWORK, PROTOCOL_VERSION, ret, 0);
+    ss << header << x;
+    return ret;
+}
+
+/** An structure for PSBTs which contain per input information */
+struct PartiallySignedInput
+{
+    CTransactionRef non_witness_utxo;
+    CTxOut witness_utxo;
+    std::map<CPubKey, std::vector<unsigned char>> partial_sigs;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+    int sighash_type = 0;
+    uint64_t index;
+    bool use_in_index = false;
+
+    PartiallySignedInput(): non_witness_utxo(), witness_utxo(), partial_sigs(), unknown() {}
+    void SetNull();
+    bool IsNull();
+};
+
+/** A version of CTransaction with the PSBT format*/
+struct PartiallySignedTransaction
+{
+    CMutableTransaction tx;
+    std::map<uint160, CScript> redeem_scripts;
+    std::map<uint256, CScript> witness_scripts;
+    std::vector<PartiallySignedInput> inputs;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+    std::map<CPubKey, std::vector<uint32_t>> hd_keypaths;
+    uint64_t num_ins = 0;
+    bool use_in_index = false;
+
+    PartiallySignedTransaction();
+    PartiallySignedTransaction(CMutableTransaction& tx, std::map<uint160, CScript>& redeem_scripts, std::map<uint256, CScript>& witness_scripts, std::vector<PartiallySignedInput>& inputs);
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+
+        // magic bytes
+        s << PSBT_MAGIC_BYTES << (char)0xff; // psbt 0xff
+
+        // Write transaction if it exists
+        if (!CTransaction(tx).IsNull()) {
+            // unsigned tx flag
+            WriteCompactSize(s, sizeof(PSBT_UNSIGNED_TX_NON_WITNESS_UTXO));
+            s << PSBT_UNSIGNED_TX_NON_WITNESS_UTXO;
+
+            // Write serialized tx to a stream
+            WriteCompactSize(s, ::GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION));
+            s << tx;
+        }
+
+        // Write redeem scripts and witness scripts
+        for (auto& entry : redeem_scripts) {
+            s << SerializeToVector(PSBT_REDEEMSCRIPT_WITNESS_UTXO, entry.first);
+            WriteCompactSize(s, entry.second.size());
+            s << MakeSpan(entry.second);
+        }
+        for (auto& entry : witness_scripts) {
+            s << SerializeToVector(PSBT_WITNESSSCRIPT_PARTIAL_SIG, entry.first);
+            WriteCompactSize(s, entry.second.size());
+            s << MakeSpan(entry.second);
+        }
+        // Write any hd keypaths
+        for (auto keypath_pair : hd_keypaths) {
+            s << SerializeToVector(PSBT_BIP32_KEYPATH_SIGHASH, Span<const unsigned char>(keypath_pair.first.begin(), keypath_pair.first.size()));
+            WriteCompactSize(s, keypath_pair.second.size() * sizeof(uint32_t));
```

Comment:
This does not work because it will be prefixed with the number of items in the keypath, not with the total size in bytes that that will be. Each item in the keypath is a uint32_t, not an unsigned char.